### ISSUE TITLE
Does it support the model wrapped by DistributedDataParallel？

### ISSUE THREAD

**Issue by guozhiyao** [tristandeleu/pytorch-meta#100]

**Comment by tristandeleu** [comment]
Currently MetaModules are likely not compatible with DistributedDataParallel unfortunately. One option you can look into is the [`DataParallel`](https://github.com/tristandeleu/pytorch-meta/blob/74cbc8ae625d85c6b954aad159ccb26b523b2240/torchmeta/modules/parallel.py) wrapper provided in Torchmeta as a drop-in replacement of `nn.DataParallel`. As I understand, DDP is supposed to replace DataParallel, but I haven't come around to get a wrapper for it in Torchmeta. However contributions for adding such a wrapper would be very welcome!

**Comment by brando90** [comment]
> Currently MetaModules are likely not compatible with DistributedDataParallel unfortunately. One option you can look into is the [`DataParallel`](https://github.com/tristandeleu/pytorch-meta/blob/74cbc8ae625d85c6b954aad159ccb26b523b2240/torchmeta/modules/parallel.py) wrapper provided in Torchmeta as a drop-in replacement of `nn.DataParallel`. As I understand, DDP is supposed to replace DataParallel, but I haven't come around to get a wrapper for it in Torchmeta. However contributions for adding such a wrapper would be very welcome!

are the modules and data loader compatible with the normal one though i.e. with `from torch.nn import DataParallel`?

**Comment by guozhiyao** [comment]
> Currently MetaModules are likely not compatible with DistributedDataParallel unfortunately. One option you can look into is the [`DataParallel`](https://github.com/tristandeleu/pytorch-meta/blob/74cbc8ae625d85c6b954aad159ccb26b523b2240/torchmeta/modules/parallel.py) wrapper provided in Torchmeta as a drop-in replacement of `nn.DataParallel`. As I understand, DDP is supposed to replace DataParallel, but I haven't come around to get a wrapper for it in Torchmeta. However contributions for adding such a wrapper would be very welcome!

I've test the modules wrapped by DDP, and the second-order derivative seems correct. The DP need to broadcast the model parameters to each devices in each forward, so you wrape the DP. But the DDP broadcast the model parameters once when the model was wrapped with DDP, and DDP will synchronize the gradients in each devices, which does not need to broadcast the parameters anymore. Does it means the torchmeta can be wrapped with DDP?

**Comment by brando90** [comment]
> > Currently MetaModules are likely not compatible with DistributedDataParallel unfortunately. One option you can look into is the [`DataParallel`](https://github.com/tristandeleu/pytorch-meta/blob/74cbc8ae625d85c6b954aad159ccb26b523b2240/torchmeta/modules/parallel.py) wrapper provided in Torchmeta as a drop-in replacement of `nn.DataParallel`. As I understand, DDP is supposed to replace DataParallel, but I haven't come around to get a wrapper for it in Torchmeta. However contributions for adding such a wrapper would be very welcome!
> 
> I've test the modules wrapped by DDP, and the second-order derivative seems correct. The DP need to broadcast the model parameters to each devices in each forward, so you wrape the DP. But the DDP broadcast the model parameters once when the model was wrapped with DDP, and DDP will synchronize the gradients in each devices, which does not need to broadcast the parameters anymore. Does it means the torchmeta can be wrapped with DDP?

@guozhiyao I think a good test is to do what you are doing. Use DDP and print the gradients to test if the synchronization are right when you are doing the forward pass of maml for example. 

Perhaps you can share your code or something to take a look at it and help, otherwise it's hard. If you have one with higher that is even better!

**Comment by brando90** [comment]
related issue: https://github.com/facebookresearch/higher/issues/98

**Comment by brando90** [comment]
> Currently MetaModules are likely not compatible with DistributedDataParallel unfortunately. One option you can look into is the [`DataParallel`](https://github.com/tristandeleu/pytorch-meta/blob/74cbc8ae625d85c6b954aad159ccb26b523b2240/torchmeta/modules/parallel.py) wrapper provided in Torchmeta as a drop-in replacement of `nn.DataParallel`. As I understand, DDP is supposed to replace DataParallel, but I haven't come around to get a wrapper for it in Torchmeta. However contributions for adding such a wrapper would be very welcome!

another option that is very likely to work for data parallelism is: https://pytorch.org/docs/stable/rpc.html#distributed-autograd-framework (yea I agree that diving your dataloaders the DistributedSampler is likely to not work. DDP is complicated and sort of magical so I'd be surprised if it did work).

The above was suggested to me in this discussion: https://discuss.pytorch.org/t/how-to-parallelize-a-loop-over-the-samples-of-a-batch/32698/7

**Comment by brando90** [comment]
closing to make it less confusing. Answer is:

1. torchmeta data loaders should be compatible with DDP
2. the meta-modules in torchmeta no.
3. higher no.

https://github.com/tristandeleu/pytorch-meta/issues/116
